### PR TITLE
feat(annotations): Pass Annotations to helper pod from chaos pod

### PR DESCRIPTION
**Pull request by uditgaurav**
Signed-off-by: Udit Gaurav <udit.gaurav@mayadata.io>

#### Issues:
- fixes: https://github.com/litmuschaos/litmus/issues/1973

#### Details:
- This PR is for passing the annotation from chaos pod to helper pod.

### CODE REVIEW

File: pkg/generic/node-memory-hog/environment/environment.go [hunk 1]
```diff
@@ -28,6 +28,7 @@ func GetENV(experimentDetails *experimentTypes.ExperimentDetails, expName string
 	experimentDetails.AuxiliaryAppInfo = Getenv("AUXILIARY_APPINFO", "")
 	experimentDetails.Delay, _ = strconv.Atoi(Getenv("STATUS_CHECK_DELAY", "2"))
 	experimentDetails.Timeout, _ = strconv.Atoi(Getenv("STATUS_CHECK_TIMEOUT", "180"))
+	experimentDetails.Annotations = map[string]string{}
```
Comment (ksatchit) [inline pkg/generic/node-memory-hog/environment/environment.go:31]: Why do we need this in environment.go?

File: pkg/generic/node-memory-hog/types/types.go [hunk 2]
```diff
@@ -24,4 +24,5 @@ type ExperimentDetails struct {
 	AuxiliaryAppInfo string
 	Timeout          int
 	Delay            int
+	Annotations      map[string]string
```
Comment (ksatchit) [inline pkg/generic/node-memory-hog/types/types.go:27]: should this be a list?

File: pkg/utils/common/common.go [hunk 3]
```diff
@@ -98,3 +98,13 @@ func GetPodAndNodeName(namespace, targetPod, appLabels string, clients clients.C
...
 }
+
+// GetChaosPodAnnotation will return the annotation on chaos pod
+func GetChaosPodAnnotation(podName, namespace string, clients clients.ClientSets) (map[string]string, error) {
+
+	pod, err := clients.KubeClient.CoreV1().Pods(namespace).Get(podName, v1.GetOptions{})
+	if err != nil {
+		return nil, errors.Errorf("fail to get the chaos pod annotation, due to %v", err)
+	}
+	return pod.Annotations, nil
```
Comment (ksatchit) [inline pkg/utils/common/common.go:109]: This is going to contain the annotations specified by user + annotations inserted at runtime by kubernetes?

Comment (ksatchit) [reply pkg/utils/common/common.go:109]: Can we get the -o yaml output comparison b/w the experiment and helper pod once the helper is launched successfully w/ the annotations?

**Comment by uditgaurav** [comment]
### Chaos Engine:

```yaml
apiVersion: litmuschaos.io/v1alpha1
kind: ChaosEngine
metadata:
  name: nginx-chaos
  namespace: default
spec:
  appinfo:
    appns: 'default'
    applabel: 'run=nginx'
    appkind: 'deployment'
  annotationCheck: 'false'
  engineState: 'active'
  #ex. values: ns1:name=percona,ns2:run=nginx
  auxiliaryAppInfo: ''
  chaosServiceAccount: pod-delete-sa
  monitoring: false
  jobCleanUpPolicy: 'retain'
  components:
    runner:
      runnerannotation:
        testrunner.io/chaos: "true"  
  experiments:
    - name: node-memory-hog
      spec:
        components:
          experimentannotation:
            testexp.io/chaos: "true"
          env:
            # set chaos duration (in sec) as desired
            - name: TOTAL_CHAOS_DURATION
              value: '30'
            # set chaos interval (in sec) as desired
            - name: MEMORY_PERCENTAGE
              value: '80'
```
### Chaos Pod

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/psp: eks.privileged
    testexp.io/chaos: "true"
  creationTimestamp: "2020-09-04T12:55:31Z"
  generateName: node-memory-hog-xqw5kw-
  labels:
    chaosUID: af92dfc5-fc5b-4964-a4ee-5e697ffb9ded
    controller-uid: 3b9fed41-4c7b-4bf8-8cac-9070e8eec255
    job-name: node-memory-hog-xqw5kw
    name: node-memory-hog
  name: node-memory-hog-xqw5kw-tjfff
  namespace: default
  ownerReferences:
  - apiVersion: batch/v1
    blockOwnerDeletion: true
    controller: true
    kind: Job
    name: node-memory-hog-xqw5kw
    uid: 3b9fed41-4c7b-4bf8-8cac-9070e8eec255
  resourceVersion: "16076836"
  selfLink: /api/v1/namespaces/default/pods/node-memory-hog-xqw5kw-tjfff
  uid: 629c81b7-d10e-4145-b14b-45b903055381
spec:
  containers:
```

### Helper Pod

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/psp: eks.privileged
    testexp.io/chaos: "true"
  creationTimestamp: "2020-09-04T12:55:36Z"
  labels:
    app: node-memory-hog
    chaosUID: af92dfc5-fc5b-4964-a4ee-5e697ffb9ded
    name: node-memory-hog-iegnct
  name: node-memory-hog-iegnct
  namespace: default
  resourceVersion: "16076916"
  selfLink: /api/v1/namespaces/default/pods/node-memory-hog-iegnct
  uid: 5514cfb9-d490-4d9f-b081-6adb9556b472
spec:
  containers:
  - args:
    - --vm
    - "1"
    - --vm-bytes
    - 80%
    - --timeout
    - "30"
    command:
    - /stress-ng
    image: litmuschaos/go-runner:latest
    imagePullPolicy: Always
    name: node-memory-hog
```

cc: @ksatchit

File: pkg/generic/node-memory-hog/environment/environment.go (hunk 1 above)
Comment (uditgaurav) [reply pkg/generic/node-memory-hog/environment/environment.go:31]: Using to save choas pod annotation (in `experimentDetails.Annotations`) and then applying it to helper pod. It is of type map[string]string.

File: pkg/utils/common/common.go (hunk 3 above)
Comment (uditgaurav) [reply pkg/utils/common/common.go:109]: https://github.com/litmuschaos/litmus-go/pull/105#issuecomment-687128278

Comment (ksatchit) [reply pkg/utils/common/common.go:109]: Can we check this one minikube or any other cluster? Interested in seeing if dynamic info - something like this: Does it use same timestamps/values as having been derived from chaos pod OR if K8s overwrites w/ its own latest runtime values? 

```
kubernetes.io/config.seen="2020-09-04T12:00:04.197547545Z"
kubernetes.io/config.source="api"
kubernetes.io/limit-ranger="LimitRanger plugin set: cpu request for container nginx"
```

File: pkg/generic/node-memory-hog/environment/environment.go (hunk 1 above)
Comment (ksatchit) [reply pkg/generic/node-memory-hog/environment/environment.go:31]: Think this is sufficient if added into types.go

Comment (ispeakc0de) [reply pkg/generic/node-memory-hog/environment/environment.go:31]: The initialization of this variable is not needed, we can directly use it.

Comment (uditgaurav) [reply pkg/generic/node-memory-hog/environment/environment.go]: yes yes it is not required!

File: pkg/utils/common/common.go (hunk 3 above)
Comment (uditgaurav) [reply pkg/utils/common/common.go:109]: Tested it with microk8s (same as minikube) the result was similar.
#### Helper pod
```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    cni.projectcalico.org/podIP: 10.1.243.198/32
    cni.projectcalico.org/podIPs: 10.1.243.198/32
    testexp.io/chaos: "true"
  creationTimestamp: "2020-09-04T18:25:35Z"
  labels:
    app: node-memory-hog
    chaosUID: a8074ed0-fdba-4e64-9060-334569fa17a9
    name: node-memory-hog-simzvs
  managedFields:
  - apiVersion: v1
    fieldsType: FieldsV1
    fieldsV1:
```

**Review by ispeakc0de — APPROVED** [review]

File: pkg/utils/common/common.go (hunk 3 above)
Comment (ksatchit) [reply pkg/utils/common/common.go:109]: Thanks, this helps!

**Review by ksatchit — APPROVED** [review]

**Pull request merged**